### PR TITLE
Roll back CommitStateDB after failing to execute handler in evm module

### LINKED ISSUE #668: Bad data may be generated due to insufficient gas during the execution of evm transactions
__System info:__ [Include Ethermint commit, operating system name, and other relevant details]

* development

__Steps to reproduce:__

1. An evm transaction is executed, and the evm.Create or evm.Call method is successfully executed, the "storage" data and "code" data in CommitStateDB have been modified.
2. Start to execute the function "csdb.Finalise". The function needs to set data to the Store (Keeper). Due to insufficient gas, panic is triggered and the transaction execution fails.
3. The data set to Store (Keeper) will be rolled back, but the "storage" data and "code" data from CommitStateDB will remain in CommitStateDB.
4.Then, the problem encountered is Similar to [issue667](https://github.com/cosmos/ethermint/issues/667)，[issue669](https://github.com/cosmos/ethermint/issues/669)

### PR DESCRIPTION
**Pull request by summerpro**
Closes: #668 #669 

## Description

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

**Comment by summerpro** [comment]
This solution can only solve #668 and #669, but cannot solve #667. The final solution needs to be discussed.

**Comment by fedekunze** [comment]
It's ok for me to use this as a short term approach. I think we might want to set everything to store instead of keeping "dirty" arrays/maps. This would be a large refactor of the evm module but it also will increase the performance as we won't need to iterate over the dirty entries during `EndBlock`

File: x/evm/handler.go [hunk 1]
```diff
@@ -14,16 +14,27 @@ import (
...
 func NewHandler(k Keeper) sdk.Handler {
-	return func(ctx sdk.Context, msg sdk.Msg) (*sdk.Result, error) {
+	return func(ctx sdk.Context, msg sdk.Msg) (result *sdk.Result, err error) {
+		snapshotStateDB := k.CommitStateDB.Copy()
+		defer func() {
+			if r := recover(); r != nil {
+				k.CommitStateDB = snapshotStateDB
+				panic(r)
+			}
+		}()
```
Comment (fedekunze) [inline x/evm/handler.go]: this requires a large comment to explain what's happening

**Review by fedekunze — COMMENTED** [review]
Thanks @summerpro! Could you add a test case to ensure this is not broken in the future?

**Comment by summerpro** [comment]
> It's ok for me to use this as a short term approach. I think we might want to set everything to store instead of keeping "dirty" arrays/maps. This would be a large refactor of the evm module but it also will increase the performance as we won't need to iterate over the dirty entries during `EndBlock`

Yes, I agree with you, the refactored solution should be able to solve #667 #668 #669

**Comment by summerpro** [comment]
> Thanks @summerpro! Could you add a test case to ensure this is not broken in the future?

You're welcome. Ok i will add it tomorrow

File: x/evm/handler.go (hunk 1 above)
Comment (summerpro) [reply x/evm/handler.go]: ok

**Comment by summerpro** [comment]
@fedekunze Can this pr be merged?

**Review by fedekunze — APPROVED** [review]

**Pull request merged**